code/lora/lora.py

```diff
@@ -0,0 +1,298 @@
+            rx_log.write(payload)
+            rx_log.write("-"*(UP_PACKET_LENGTH-len(payload)))
+            rx_log.write("\n")
+        print("RX:", payload)
+        self.ack[self.last_tx_mode] = True #Acknowledge the previous packet
+
+
+    def append_mgblk_buffer(self):
+        """
+        Adds new entries of casual input logs to the buffer queue.
+        In case of an overflow, destroyes some data at hte front of the buffer.
+        """
+
+        if len(self.buffer) > MAX_BUFFER_LENGTH:
+            print("Buffer overflow")
+            for _ in range(MAX_BUFFER_APPEND):
+                self.buffer.pop(0) #Destroy enough data for the next reading to fit
+
+        for path in TX_FILES:
+            with open(path, 'r') as tx:
+                tx.seek(0, 2)
+                pos = tx.tell() #File length
+                diff = round((pos-self.filepos[path])/DOWN_PACKET_LENGTH) #Amount of new readings
+                self.filepos[path] = pos
+                tx.seek(pos-MAX_BACKREAD*TX_LINE_LENGTH, 0)
+                lines = [tx.read(TX_LINE_LENGTH) for i in range(MAX_BACKREAD)] #Read [MAX_BACKREAD] lines
+            waiting = min(MAX_BACKREAD, diff)
+            for i in range(MAX_BACKREAD-waiting, MAX_BACKREAD): #Append only the new ones to buffer
+                self.buffer.append(HEADER.MSGPACK + lines[i])
+
+
+    def tx_default(self):
+        """
+        Retrieves the first packet from the buffer
+        and transmits it.
+        """
+
+        if len(self.buffer) == 0:
+            print("Buffer empty!")
+            self.buffer.insert(0, "SSBuffer Empty!")
+
+        payload_str = self.buffer.pop(0)
+        payload_str += "-"*(DOWN_PACKET_LENGTH-len(payload_str))
+
+        payload_b = list(bytearray(payload_str, "utf-8"))
+        payload = HEADER.LORA + payload_b
+        print("TX: " + payload_str)
+
+        self.write_payload(payload)
+        self.set_mode(MODE.TX)
+        time.sleep(TX_TIME)
+        self.reset_ptr_rx()
+        self.set_mode(MODE.RXCONT)
+        self.last_tx_mode = 'default'
+
+
+    def tx_image(self):
+        """
+        Transmits a chunk of picamera image data.
+        If no ACK was received, same payload is sent.
+        Otherwise a new chunk is loaded.
+        Upon reaching end of file, stops the transmission.
+        """
+
+        if self.ack['image']: #If ACK received, send next slice of data
+            self.image_waiting = [HEADER.IMAGE_B] + list(self.image.read(DOWN_PACKET_LENGTH-1))
+            if self.image_waiting == [HEADER.IMAGE_B]: #If EOF, end image transmission
+                self.sending_image = False
+                return
+
+        self.ack['image'] = False
+        packet_length = len(self.image_waiting)
+        if packet_length < DOWN_PACKET_LENGTH:
+            self.image_waiting += (DOWN_PACKET_LENGTH-packet_length)*[45]
+        print("TX:", self.image_waiting)
+
+        self.write_payload(HEADER.LORA + self.image_waiting)
```

Comment:
that part looks like duplicate from various places, I would consider making a function with `payload` parameter or whatever that does that

Reply:
done just that